packages/js/core/src/types/WrapError.ts

```diff
@@ -0,0 +1,274 @@
+import { CleanResolutionStep } from "../algorithms";
+
+export type ErrorSource = Readonly<{
+  file?: string;
+  row?: number;
+  col?: number;
+}>;
+
+/**
+Wrap error codes provide additional context to WrapErrors.
+
+Error code naming convention (approximate):
+    type of handler
+    type of functionality
+    piece of functionality
+    ==> handler_typeFn_pieceFn
+
+Error code map:
+  0-24 -> Client
+  25-49 -> URI resolution
+  50-74 -> Wrapper invocation & sub-invocation
+  75-255 -> Unallocated
+ */
+export enum WrapErrorCode {
+  UNKNOWN,
+  CLIENT_LOAD_WRAPPER_ERROR,
+  CLIENT_GET_FILE_ERROR,
+  CLIENT_GET_IMPLEMENTATIONS_ERROR,
+  CLIENT_VALIDATE_RESOLUTION_FAIL,
+  CLIENT_VALIDATE_ABI_FAIL,
+  CLIENT_VALIDATE_RECURSIVE_FAIL,
+  URI_RESOLUTION_ERROR = 25,
+  URI_RESOLVER_ERROR,
+  URI_NOT_FOUND,
+  WRAPPER_INVOKE_ABORTED = 50,
+  WRAPPER_SUBINVOKE_ABORTED,
+  WRAPPER_INVOKE_FAIL,
+  WRAPPER_READ_FAIL,
+  WRAPPER_INTERNAL_ERROR,
+  WRAPPER_METHOD_NOT_FOUND,
+  WRAPPER_ARGS_MALFORMED,
+}
+
+export interface WrapErrorOptions {
+  code: WrapErrorCode;
+  uri: string;
+  method?: string;
+  args?: string;
+  source?: ErrorSource;
+  resolutionStack?: CleanResolutionStep;
+  cause?: unknown;
+  prev?: Error;
+}
+
+type RegExpGroups<T extends string> =
+  | (RegExpExecArray & {
+      groups?: { [name in T]: string | undefined } | { [key: string]: string };
+    })
+  | null;
+
+export class WrapError extends Error {
+  readonly name: string = "WrapError";
+  readonly code: WrapErrorCode;
+  readonly reason: string;
+  readonly uri: string;
+  readonly method?: string;
+  readonly args?: string;
+  readonly source?: ErrorSource;
+  readonly resolutionStack?: CleanResolutionStep;
+  readonly cause?: unknown;
```

Comment:
Since this is essentially a native Error, and WrapError has a need to be serialized, I think `cause` should be of type `string` (Serialized Error basically).
I'm also open to any other solution, but I don't think we should have an `unknown` type in WrapError

Reply:
It is necessary to have cause be of `unknown` type to accommodate uri resolver errors, which are untyped. WrapError provides an incremental improvement over the status quo in this respect.

Consider the status quo: a uri resolver error (of type `unknown`) that is thrown during a sub-invocation is completely lost. It is ignored when the error is stringified. The user loses all of the information.

In `WrapError`, the uri resolver error is itself stringified and included in the error message. While we lose the uri resolver error itself, we at least keep a stringified version.

In both the status quo and the current PR, the full uri resolver error of `unknown` type is still attached to the error object when it occurs during a top-level invocation. If we change `cause` to be of type `string`, we lose this.

I know you thought before that uri resolver errors should be of type `unknown` to provide flexibility to custom uri resolvers. If `cause` becomes a `string`, there is little point to uri resolver errors being of any type but `string` since that's all the user will ever get.

When serializing `WrapError`, I 100% agree the value of `cause` should be a `string`.

What do you think?

Reply:
Sure, I'm down to keep it `unknown` for now.